src/vs/workbench/parts/explorers/browser/treeExplorerViewlet.ts

```diff
@@ -0,0 +1,88 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import { TPromise } from 'vs/base/common/winjs.base';
+import { Builder, Dimension } from 'vs/base/browser/builder';
+import { Orientation } from 'vs/base/browser/ui/splitview/splitview';
+import { IAction } from 'vs/base/common/actions';
+import { IViewletView, Viewlet } from 'vs/workbench/browser/viewlet';
+import { ITelemetryService } from 'vs/platform/telemetry/common/telemetry';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { TreeExplorerView } from 'vs/workbench/parts/explorers/browser/views/treeExplorerView';
+import { TreeExplorerViewletState } from 'vs/workbench/parts/explorers/browser/views/treeExplorerViewer';
+import { IActivityService } from 'vs/workbench/services/activity/common/activityService';
+import { VIEWLET_ID_ROOT } from 'vs/workbench/parts/explorers/common/treeExplorer';
+
+export class TreeExplorerViewlet extends Viewlet {
+	private static _idCounter = 1;
+
+	private viewletContainer: Builder;
+	private view: IViewletView;
+
+	private viewletState: TreeExplorerViewletState;
+
+	private externalViewletId: string;
+	private treeNodeProviderId: string;
+
+	constructor(
+		@ITelemetryService telemetryService: ITelemetryService,
+		@IInstantiationService private instantiationService: IInstantiationService,
+		@IActivityService private activityService: IActivityService
+	) {
+		super(VIEWLET_ID_ROOT + TreeExplorerViewlet._idCounter, telemetryService);
```

Comment:
@octref this is very confusing and fragile. the viewlet ID is now suddenly something dynamic and not static? and then you get an external ID from somewhere and return it from `getId()`? All of this has to change :-/